src/qt/veil/miningwidget.cpp

```diff
@@ -0,0 +1,147 @@
+    mainWindow(walletView)
+{
+    ui->setupUi(this);
+    ui->title->setProperty("cssClass" , "title");
+
+    // SET THE NUMBER OF THREADS AND ALGORITHM HERE
+    mineOn = GenerateActive();
+
+    ui->cmbAlgoSelect->addItems({"RandomX","ProgPow","SHA256D"});
+    ui->cmbAlgoSelect->setCurrentIndex(GetMiningAlgorithm());
+
+    ui->lblMaxThreadsAvailable->setText(QString::number(GetNumCores() - 1));
+    
+
+    connect(ui->btnUpdateAlgo, SIGNAL(clicked()), this, SLOT(onUpdateAlgorithm()));
+    connect(ui->btnAllThreads, SIGNAL(clicked()), this, SLOT(onUseMaxThreads()));
+    connect(ui->btnActiveMine, SIGNAL(clicked()), this, SLOT(onToggleMiningActive()));
+
+}
+
+void MiningWidget::setMineActiveTxt(bool mineActive){
+    ui->lblMineActive->setText(mineActive ? "Mining Active" : "Mining Stopped");
+    ui->lblMineActive->setStyleSheet(mineActive ? "QLabel{color:green;}" : "QLabel{color:red;}");
+    ui->btnActiveMine->setText(mineActive ? "Stop Mining" : "Start Mining");
+}
+
+void MiningWidget::setWalletModel(WalletModel *model){
+    this->walletModel = model;
+
+    connect(model, SIGNAL(updateMiningFields()), this, SLOT(updateMiningFields()));
+}
+
+MiningWidget::~MiningWidget()
+{
+    delete ui;
+}
+
+void MiningWidget::onUpdateAlgorithm() {
+    if (mineOn) {
+        openToastDialog(QString::fromStdString("Mining must be stopped to change algorithm!"), mainWindow->getGUI());
+    } else {
+        currentMiningAlgo = ui->cmbAlgoSelect->currentIndex(); 
+        std::string algoStr = GetMiningType(currentMiningAlgo, false, false);
+        bool setAlgoResult = SetMiningAlgorithm(algoStr); 
+
+        if (setAlgoResult) {
+            openToastDialog(QString::fromStdString("Aglorithm Switch Success!"), mainWindow->getGUI());
+            ui->lblCurrentAlgo->setText(QString::fromStdString(algoStr));
+        } else {
+            openToastDialog(QString::fromStdString("Algorithm Switch Failed!"), mainWindow->getGUI());
+        }
+
+        setThreadSelectionValues(currentMiningAlgo);
+    }
+}
+
+void MiningWidget::onToggleMiningActive() {
+
+    // need to define "coinbaseScript"
+    std::shared_ptr<CReserveScript> coinbase_script;
+    walletModel->wallet().getWalletPointer()->GetScriptForMining(coinbase_script);
+
+    //throw an error if no script was provided
+    if (!coinbase_script || coinbase_script->reserveScript.empty()) {
+       openToastDialog("No coinbase script available", mainWindow->getGUI());
+       return;
+    }
+
+    int nAlgo = GetMiningAlgorithm();
+    int nCores = GetNumCores();
+    int nThreads = ui->numThreads->text().toInt();
+
+    if (!mineOn) {
+        if (nThreads > (nCores - 1)) {
+            openToastDialog("Too many threads", mainWindow->getGUI());
+            return;
+        }
+            
+        if ((nAlgo == MINE_SHA256D) && (nThreads > (nCores - 1)))
+            openToastDialog(QString::fromStdString("SHA256D limited to " + (nCores-1)), mainWindow->getGUI());
```

Comment:
Two comments here.  First:  I'm contemplating it, but for cli based mining, they -can- override this check; so I'm leaning towards a "do it anyway" type button in case they really want to do it.

Second:  The second "if" statement won't be hit.  If `(nThreads > (nCores - 1))` then it will pop "Too many threads" and won't ever get down to the check about the SHA256D limitation.  The only way it can get to the MIN_SHA256D check is if the second condition isn't met.    Quite simply, that first nThreads check shouldn't be there, since it is intended to restrict SHA256D (Per PR #879)

Reply:
We can implement an override in a later iteration.  This PR can lay down a baseline.  I'm inclined to leave the first check and remove the 2nd check (first check applies to ProgPow as well).  It is almost irrelevant as the GUI currently restricts the user.  I'm inclined to leave it for security check.  Thoughts?

Reply:
I'm fine with the first as a safety check (until the option to override is added), given that the GUI should prevent the casual user from hitting this.

Reply:
With a better understanding of the history of the checks I'm OK with removing the first one.  I'd like to do something where if the user exceeds the recommended number of threads (which would be Cores -1) a popup appears allowing the user to override.  I'm still inclined to have that be a separate PR.

I have the formatting changes ready but have not pushed them.  @CaveSpectre11, let me know if you want the thread limits changed in this PR or a different one.  After that I will either push the changes I have now or add the other changes and push it in one go.